utils/lazy-pages-fuzzer/src/generate/globals.rs

```diff
@@ -0,0 +1,201 @@
+                        .elements_mut()
+                        .insert(insert_at_pos, instr.clone());
+                }
+            }
+        }
+
+        // Add global exports
+        let mut builder = builder::from_module(module);
+        for global in global_names.iter() {
+            builder.push_export(
+                builder::export()
+                    .field(global)
+                    .internal()
+                    .global(next_global_idx)
+                    .build(),
+            );
+            builder.push_global(
+                builder::global()
+                    .mutable()
+                    .value_type()
+                    .i64()
+                    .init_expr(Instruction::I64Const(0))
+                    .build(),
+            );
+
+            next_global_idx += 1;
+        }
+
+        Ok((builder.build(), self.unstructured))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use gear_wasm_instrument::parity_wasm::elements::Internal;
+
+    use super::*;
+
+    const TEST_PROGRAM_WAT: &str = r#"
+        (module
+            (func (export "main") (result i32)
+                i32.const 42
+            )
+        )
+    "#;
+
+    #[test]
+    fn test_inject_globals() {
+        let unstructured = Unstructured::new(&[0u8; 32]);
+        let config = InjectGlobalsConfig {
+            max_global_number: 3,
+            max_access_per_func: 3,
+        };
+        let globals = InjectGlobals::new(unstructured, config);
+
+        let wasm = wat::parse_str(TEST_PROGRAM_WAT).unwrap();
+        let module = Module::from_bytes(wasm).unwrap();
+        let (module, _) = globals.inject(module).unwrap();
+
+        assert_eq!(module.globals_space(), 3);
+        assert_eq!(
+            module
+                .export_section()
+                .unwrap()
+                .entries()
+                .iter()
+                .filter(|export| { matches!(export.internal(), Internal::Global(_)) })
+                .count(),
+            3
+        );
+    }
+
+    #[test]
+    fn test_globals_modified() {
+        let unstructured = Unstructured::new(&[3u8; 32]);
+        let config = InjectGlobalsConfig {
+            max_global_number: 3,
+            max_access_per_func: 3,
+        };
+        let globals = InjectGlobals::new(unstructured, config);
+
+        let wasm = wat::parse_str(TEST_PROGRAM_WAT).unwrap();
+        let module = Module::from_bytes(wasm).unwrap();
+        let (module, _) = globals.inject(module).unwrap();
+
+        let module = sandbox_wasmi::Module::from_buffer(module.into_bytes().unwrap()).unwrap();
+        let instance =
+            sandbox_wasmi::ModuleInstance::new(&module, &sandbox_wasmi::ImportsBuilder::default())
+                .unwrap()
+                .assert_no_start();
+        let _ = instance
+            .invoke_export("main", &[], &mut sandbox_wasmi::NopExternals)
+            .unwrap();
+
+        // Assert that global was modified (initially 0)
```

Comment:
better check that in the test

Reply:
maybe you can even check the actual value after the invocation

Reply:
Fixed